staging/cse/windows/nvidiagpudriverfunc.ps1

```diff
@@ -0,0 +1,178 @@
+            }
+            else {
+                $ErrorMsg = "GPU Driver Installation Failed! Code: $($p.ExitCode)"
+                Write-Log $ErrorMsg
+                Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_FAILED -ErrorMessage $ErrorMsg
+            }
+  
+            if ($Setup.RebootNeeded -or $p.ExitCode -eq 1) {
+                Write-Log "Reboot is needed for this GPU Driver..."
+                $global:RebootNeeded = $true
+            }
+        }
+        catch [System.TimeoutException] {
+            $ErrorMsg = "Timeout $Timeout s exceeded. Stopping the installation process. Reboot for another attempt."
+            Write-Log $ErrorMsg
+            Stop-Process -InputObject $p -Force
+            Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_TIMEOUT -ErrorMessage $ErrorMsg
+        }
+        catch {
+            $ErrorMsg = "Exception: $($_.ToString())"
+            Write-Log $ErrorMsg # the status file may get over-written when the agent re-attempts this step
+            Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_FAILED -ErrorMessage $ErrorMsg
+        }
+      
+    }
+    catch {
+        $FatalError += $_
+        $errorCount = $FatalError.Count
+        $ErrorMsg = "A fatal error occurred. Number of errors: $errorCount. Error details: $($_ | Out-String)"
+        Write-Log $ErrorMsg
+        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_FAILED -ErrorMessage $ErrorMsg
+    }
+}
+  
+function Get-Setup {
+    param(
+        [Parameter(Mandatory = $true)]
+        [PSCustomObject]$DriverConfig
+    )
+
+    [OutputType([hashtable])]
+    
+    $GpuDriverURL = $DriverConfig.GpuDriverURL
+    $fileName = [IO.Path]::GetFileName($GpuDriverURL)
+      
+    if ($GpuDriverURL -eq $null) {
+        $ErrorMsg = "DriverURL is not properly specified."
+        Write-Log $ErrorMsg
+        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_URL_NOT_SET -ErrorMessage $ErrorMsg
+    }
+
+    Write-Log "gpu url is set to $GpuDriverURL"
+
+    # check if vm size is nv series. if so, set RebootNeeded to be true
+    try {
+        $Compute = Get-VmData
+        $vmSize = $Compute.vmSize
+    }
+    catch {
+        $ErrorMsg = "Failed to query the SKU information."
+        Write-Log $ErrorMsg
+        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_SKU_INFO_NOT_FOUND -ErrorMessage $ErrorMsg
+    }
+      
+    $Setup = @{
+        RebootNeeded = ($vmSize -ne $null -and $Compute.vmSize -match "_NV")
+        Target       = "$RootDir\..\$fileName"
+    }
+
+    Write-Log "Downloading from $GpuDriverURL to $($Setup.Target)"
+    DownloadFileOverHttp -Url $GpuDriverURL -DestinationPath $Setup.Target -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_DOWNLOAD_FAILURE
+  
+    return $Setup
+}
+  
+function Get-VmData {
```

Comment:
We have a function `Retry-Command`. You can reference the existing code, for example

    try {
        $response = Retry-Command -Command "Invoke-RestMethod" -Args @{Uri=$uri; Method="Get"; ContentType="application/json"; Headers=$headers} -Retries 5 -RetryDelaySeconds 10
    } catch {
        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GET_SUBNET_PREFIX -ErrorMessage "Error getting subnet prefix. Error: $_"
    }

Reply:
refactored to use existing Retry-Command